File: src/modules/MetadataBrowser/packageXml.ts
export interface PackageSelection {
  metadata_type: string;
  members: string[];
}

export function generatePackageXml(selections: PackageSelection[], apiVersion: string): string {
  if (selections.length === 0) {
    return `<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <version>${apiVersion}</version>
</Package>`;
  }

  const sorted = [...selections].sort((a, b) => a.metadata_type.localeCompare(b.metadata_type));
  const xmlParts = sorted.map(({ metadata_type, members }) => {
    const membersXml = [...members]
      .sort((a, b) => a.localeCompare(b))
      .map((m) => `        <members>${escapeXml(m)}</members>`)
      .join("\n");
    return `    <types>\n${membersXml}\n        <name>${escapeXml(metadata_type)}</name>\n    </types>`;
  });

  return `<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
${xmlParts.join("\n")}
    <version>${apiVersion}</version>
</Package>`;
}

export function parsePackageXml(xml: string): PackageSelection[] {
  const parser = new DOMParser();
  const doc = parser.parseFromString(xml, "application/xml");
  const parserError = doc.querySelector("parsererror");
  if (parserError) {
    throw new Error("XML 语法错误，请检查标签是否闭合。");
  }

  const packageNode = doc.querySelector("Package");
  if (!packageNode) {
    throw new Error("缺少 <Package> 根节点。");
  }

  const types = Array.from(doc.getElementsByTagName("types"));
  const grouped = new Map<string, Set<string>>();
  for (const typeNode of types) {
    const name = typeNode.getElementsByTagName("name")[0]?.textContent?.trim() ?? "";
    if (!name) continue;
    const members = Array.from(typeNode.getElementsByTagName("members"))
      .map((node) => node.textContent?.trim() ?? "")
      .filter(Boolean);
    if (!grouped.has(name)) grouped.set(name, new Set<string>());
    const set = grouped.get(name)!;
    for (const member of members) set.add(member);
  }

  return Array.from(grouped.entries())
    .map(([metadata_type, members]) => ({ metadata_type, members: Array.from(members).sort() }))
    .sort((a, b) => a.metadata_type.localeCompare(b.metadata_type));
}

function escapeXml(input: string): string {
  return input
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

File: src-tauri/src/metadata/package_xml.rs
use crate::metadata::models::SelectionItem;

pub fn generate_package_xml(selections: &[SelectionItem], api_version: &str) -> String {
    if selections.is_empty() {
        return format!(
            r#"<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <version>{}</version>
</Package>"#,
            api_version
        );
    }

    let mut sorted = selections.to_vec();
    sorted.sort_by(|a, b| a.metadata_type.cmp(&b.metadata_type));

    let mut out = String::from(
        r#"<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
"#,
    );

    for item in sorted {
        if item.members.is_empty() {
            continue;
        }
        let mut members = item.members.clone();
        members.sort();
        out.push_str("    <types>\n");
        for m in members {
            out.push_str("        <members>");
            out.push_str(&escape_xml(&m));
            out.push_str("</members>\n");
        }
        out.push_str("        <name>");
        out.push_str(&escape_xml(&item.metadata_type));
        out.push_str("</name>\n");
        out.push_str("    </types>\n");
    }

    out.push_str("    <version>");
    out.push_str(api_version);
    out.push_str("</version>\n</Package>");
    out
}

fn escape_xml(input: &str) -> String {
    input
        .replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('"', "&quot;")
        .replace('\'', "&apos;")
}
